[hbase098] Add support for RowFilter's and perform Deterministic Workload

ricardomacedo13 opened the pull request:
I was using this benchmark with the hbase098 client to run some tests on the project of my master thesis. I tested the four hbase operations but notice that I wasn't able to do Filters, more specifically RowFilters.
This way, I added a new operation on hbase client, being able to perform filters by row (RowFilter). It is possible to use the EQUAL, GREATER, GREATER_OR_EQUAL, LESS and LESS_OR_EQUAL operands.
I also added new properties to achieve this solution:
- define proportions for all comparison operands;
- comparevalue, value to perform the comparisons and return the filtered results;
- seed, a long to get deterministic workloads.

As referred in the title, I also added the seed property and changed the Utils.random() function, in order to achieve deterministic results, being possible to perform multiple executions and get the same results for the same seed. This way we can get more consistent measurements for and verify the behavior for the same results.

risdenk commented:
@ricardomacedo13 at first glance there are a LOT of unrelated formatting changes and a bunch of things commented out that shouldn't be (pom.xml, etc). It makes it really hard to review.

ricardomacedo13 commented:
@risdenk thank you for the feedback. I forgot to uncomment the pom.xml and other files. I'll correct this ASAP.
In the unrelated formatting changes, as you mentioned, to be able to perform successfully the deterministic workload option and the filter operation I changed some main files (such as DB, DBWrapper, HBaseClient, Utils, and so on).
As I see, the only solution to not collude with the original implementation, is to create a separate instance of the HBase098 Client and perform filter operand (e.g. maintain the original HBaseClient file and add another instance, HBaseClientWithFilter).
For the deterministic workload, I changed the Utils.random() function, adding a condition to perform the the random operation with a random or a static seed; and added the setSeed function.

ricardomacedo13 commented:
All the comments and unnecessary formating are fixed.
For the deterministic workload, I changed the Utils.random() function, allowing generators to perform deterministic workloads with static seed. This operation does not interfere with the previous versions of the platform.
For the Filter operation, I updated the Scan method, allowing to perform scan operations or filter operations.
This way, there are no conflicts with the previous versions.

risdenk commented on core/src/main/java/com/yahoo/ycsb/Utils.java:
```diff
@@ -32,15 +32,25 @@
-      rng.set(ret);
+    public static void setSeed(String s) {
+      seed = s;
+    }
```
Use null instead of "false" as the empty value?

ricardomacedo13 replied:
100% right.
I've already changed in this new commit.

risdenk commented on core/src/main/java/com/yahoo/ycsb/workloads/FilterWorkload.java:
```diff
@@ -0,0 +1,724 @@
+    seed = p.getProperty(SEED_PROPERTY, SEED_PROPERTY_DEFAULT);
+
+//    set seed property in Utils class
+    if(!seed.equals("false")) {
```
use null instead of "false" as default value?

risdenk commented on core/src/main/java/com/yahoo/ycsb/workloads/FilterWorkload.java:
```diff
@@ -0,0 +1,724 @@
+      String compOperation = doCompareOperation();
+
+//    insert a compare value to perform the comparisons in the filter
+      if (comparevalue.equals("false")) {
```
use null instead of "false" as default value?

risdenk commented on core/src/main/java/com/yahoo/ycsb/workloads/FilterWorkload.java:
```diff
@@ -0,0 +1,724 @@
+   * The default seed property. The default value of the seed is false, giving for
+   * different executions, different results.
+   */
+  public static final String SEED_PROPERTY_DEFAULT = "false";
```
use null instead of "false" as default value?

risdenk commented on core/src/main/java/com/yahoo/ycsb/workloads/FilterWorkload.java:
```diff
@@ -0,0 +1,724 @@
+  /**
+   * The default value for the compare value.
+   */
+  public static final String COMPARE_VALUE_DEFAULT = "false";
```
use null instead of "false" as default value?

ricardomacedo13 replied:
100% right.
I've already changed in this new commit.

risdenk commented on core/src/main/java/com/yahoo/ycsb/workloads/FilterWorkload.java:
```diff
@@ -0,0 +1,724 @@
+  /**
+   * Check if the scan operation is standard or filter
+   */
+  public static final String IS_FILTER_DEFAULT = "false";
```
use null instead of "false" as default value?

ricardomacedo13 replied:
100% right.
I've already changed in this new commit.

risdenk commented on core/src/main/java/com/yahoo/ycsb/workloads/FilterWorkload.java:
```diff
@@ -0,0 +1,724 @@
+  /**
+   * The default value for the start key name.
+   */
+  public static final String START_KEY_NAME_DEFAULT = "false";
```
use null instead of "false" as default value?

ricardomacedo13 replied:
100% right.
I've already changed in this new commit.

risdenk commented on hbase098/src/main/java/com/yahoo/ycsb/db/HBaseClient.java:
```diff
@@ -22,502 +22,525 @@
-    public boolean _usePageFilter = true;
-
-    public static final int HttpError=-2;
-
```
Does this work since string "false" and boolean false are different?

ricardomacedo13 replied:
This piece of code was already in the previous versions.

risdenk commented on hbase098/src/main/java/com/yahoo/ycsb/db/HBaseClient.java:
```diff
@@ -22,502 +22,525 @@
-            THREAD_COUNT.getAndIncrement();
-            synchronized(THREAD_COUNT) {
-              if (_hConn == null){
-                _hConn = HConnectionManager.createConnection(config);
```
Is this just for debugging?

ricardomacedo13 replied:
This piece of code was already in the previous versions.

risdenk commented on workloads/workload_filter:
```diff
@@ -0,0 +1,58 @@
+# Copyright (c) 2010 Yahoo! Inc. All rights reserved.
```
Wrong copyright

risdenk commented on workloads/workloada:
```diff
@@ -31,6 +31,7 @@ readallfields=true
 readproportion=0.5
 updateproportion=0.5
 scanproportion=0
+filterproportion=0
```
Does this have to be added? Can it be missing?

ricardomacedo13 replied:
Forget to remove this tag.
I already changed this in all workload files.

risdenk commented on core/src/main/java/com/yahoo/ycsb/workloads/CoreWorkload.java:
```diff
@@ -17,32 +17,12 @@
 
 package com.yahoo.ycsb.workloads;
```
Only changes to imports here for this file?

risdenk commented on core/src/main/java/com/yahoo/ycsb/GoodBadUglyDB.java:
```diff
@@ -1,12 +1,12 @@
-/**                                                                                                                                                                                
+/**
```
Only changes to imports for this file?

ricardomacedo13 replied:
Sorry about. I already changed that in my new commit.
My IDE had import optimization enabled.

risdenk commented on hbase098/src/main/java/com/yahoo/ycsb/db/HBaseClient.java:
```diff
@@ -22,502 +22,525 @@
 import com.yahoo.ycsb.DBException;
```
Need to update the copyright for this file

ricardomacedo13 replied:
Copyright updated.

risdenk commented on core/src/main/java/com/yahoo/ycsb/workloads/FilterWorkload.java:
```diff
@@ -0,0 +1,724 @@
+package com.yahoo.ycsb.workloads;
```
Missing license header
